client/protocol.go

```diff
@@ -0,0 +1,422 @@
+package client
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	curpapi "github.com/xline-kv/go-xline/api/curp"
+	xlineapi "github.com/xline-kv/go-xline/api/xline"
+	"github.com/xline-kv/go-xline/xlog"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+// Protocol client
+type client struct {
+	// Current leader
+	leader uint64
+	// Inner protocol clients
+	inner []curpapi.ProtocolClient
+	// All servers' `Connect`
+	connects map[uint64]string
+	// Curp client timeout settings
+	timeout clientTimeout
+}
+
+// Build client from addresses, this method will fetch all members from servers
+func BuildCurpClientFromAddrs(addrs []string) (*client, error) {
+	logger := xlog.GetLogger()
+	clientTimeout := NewDefaultClientTimeout()
+
+	var clients []curpapi.ProtocolClient
+
+	for _, addr := range addrs {
+		conn, err := grpc.Dial(addr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithIdleTimeout(clientTimeout.idleTimeout))
+		if err != nil {
+			logger.Warn("failed to dial grpc", zap.Error(err), zap.String("addr", addr))
+		} else {
+			protocolClient := curpapi.NewProtocolClient(conn)
+			clients = append(clients[:], protocolClient)
+		}
+	}
+
+	cluster, err := fetchCluster(clients, clientTimeout.proposeTimeout)
+	if err != nil {
+		return nil, err
+	}
+
+	return &client{
+		inner:    clients,
+		connects: cluster.AllMembers,
+		timeout:  clientTimeout,
+	}, nil
+}
+
+type fetchClusterRes struct {
+	res *curpapi.FetchClusterResponse
+	err error
+}
+
+// Fetch cluster from server
+func fetchCluster(protocolClients []curpapi.ProtocolClient, proposeTimeout time.Duration) (*curpapi.FetchClusterResponse, error) {
+	fetchClusterRespCh := make(chan fetchClusterRes)
+
+	for _, protocolClient := range protocolClients {
+		client := protocolClient
+		go func() {
+			ctx, cancel := context.WithTimeout(context.Background(), proposeTimeout)
+			defer cancel()
+			res, err := client.FetchCluster(ctx, &curpapi.FetchClusterRequest{})
+			fetchClusterRespCh <- fetchClusterRes{res: res, err: err}
+		}()
+	}
+
+	for i := 0; i < len(protocolClients); i++ {
+		fetchClusterRes := <-fetchClusterRespCh
+
+		if fetchClusterRes.err != nil {
+			continue
+		} else {
+			return fetchClusterRes.res, nil
+		}
```

Comment:
return when `fetchClusterRes.err == nil`, we don't need an else branch here.

Reply:
modified